mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala

```diff
@@ -334,10 +334,10 @@ class KMeans @Since("1.5.0") (
     val summary = new KMeansSummary(
       model.transform(dataset), $(predictionCol), $(featuresCol), $(k))
     model.setSummary(Some(summary))
-    instr.logSuccess(model)
     if (handlePersistence) {
       instances.unpersist()
     }
+    instr.logSuccess(model)
```

Comment:
The `handlePersistence` check in `KMeans` at L309 should also be updated to use `dataset.storageLevel`. Since we're touching KMeans here anyway we may as well do it now.